helm-charts/auth-apisix/README.md

```diff
@@ -0,0 +1,87 @@
+
+2. Run keycloak, setup a realm with OIDC based authentication and add users with passwords for each user.
+
+Steps to start keycloak from official keycloak helm chart
+
+```sh
+# Prerequisite: Create a PersistentVolume of 9Gi for keycloak-postgress with RWO access (to persist updated keycloak configuration)
+# Below is a reference to create PersistentVolume which will be used by keycloak-postgress
+kubectl apply -f - <<EOF
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: data-keycloak-postgresql-0
+spec:
+  capacity:
+    storage: 9Gi
+  hostPath:
+    path: "/mnt/keycloak-vol"
+  accessModes:
+    - ReadWriteOnce
+EOF
+
+# install keycloak with helm by setting
+helm install keycloak oci://registry-1.docker.io/bitnamicharts/keycloak --version 22.1.0 --set auth.adminUser=admin --set auth.adminPassword=admin
+
+# Access keycloak UI through service/keycloak to do the necessary configurations
+```
+
+Once the keycloak pod is up and running, access the UI through the Keycloak's NodePort service and set the necessary configuration
+
+## Update values
+
+Update the following values in values.yaml
+
+1. Update all the entries in oidc config pertaining to your identity provider
+
+2. Update all the entries in API specific configs
+
+## Install
+
+```sh
+# Install apisix api gateway and ingress controller
+helm install auth-apisix apisix/apisix -f values_apisix_gw.yaml --create-namespace --namespace auth-apisix
+
+# WAIT UNTIL apisix-ingress-controller POD IS READY by checking status with 'kubectl get -n auth-apisix pods'
+# The pod is ready when READY status shows 1/1
+
+# Publish authenticated APIs in APISIX gateway
+helm install auth-apisix-crds . --namespace auth-apisix
+```
+
+## Usage
+
+The published APIs in apisix gateway are accessible through auth-apisix-gateway kubernetes service. By default, it is a NodePort service and can be accessed as:
+
+```sh
+export NODE_PORT=$(kubectl get --namespace auth-apisix -o jsonpath="{.spec.ports[0].nodePort}" services auth-apisix-gateway)
+export NODE_IP=$(kubectl get nodes --namespace auth-apisix -o jsonpath="{.items[0].status.addresses[0].address}")
+echo http://$NODE_IP:$NODE_PORT
+
+# the autherticated endpoint published in APISIX gateway can be accessed as: http://$NODE_IP:$NODE_PORT/<published endpoint uri>
+
+```
+
+</br>
+Apisix helm chart provides configs to change the service type to other options like LoadBalancer (apisix.service.type) and externalTrafficPolicy to 'local'(apisix.service.externalTrafficPolicy). These can be added in apisix-helm/values.yaml </br></br>
+While accessing the published APIs, the HTTP Authorization header of the request should contain the Access token provided by Identity provider as 'Bearer \<Access Token\>' </br></br>
+The access token, refresh token, userinfo and user roles can be obtained by invoking OIDC auth endpoint through UI or token endpoint through curl and providing user credentials. </br>
+
+## Uninstall
```

Comment:
Please also add a paragraph of how to use the APISIX gateway to cnnect to chatqna after following the above steps.

Reply:
@ckhened I would like this PR organized a little differently. 
1) APISIX is a gateway/proxy, certainly a infrastructure related. Would you kindly make a directory under GenAIInfra called Gateway and then under that add APISIX.
so GenAIInfra/Gateway/APISIX/helmchart ...

2) Any new/additional Authn&Z stuff lets put under https://github.com/opea-project/GenAIInfra/tree/main/authN-authZ. We already have Keycloak there is my understanding.

3) Under APISIX add docs on how to include authN&Z.

Please work with @Ruoyu-y to discuss restructuring. 

Thank you for addressing the other comments.

Reply:
@mkbhanda Sure. I'll sync up with @Ruoyu-y for restructuring

Reply:
How about we create 2 subdirectories under the top level authN-authZ dir, one for the current istio gateway implemenation, and one for the apisix implemenation. And also add a top level README under authN-authZ for users to reference different options? @Ruoyu-y any comment?

Reply:
@ckhened authN-authZ is originally defined for all authentication and authorization related stuff. If you are leveraging the same keycloak, we can have a directory for OIDC provider that can be used by both implementations. I suggest to make the directory structure like this:
```
authN-authZ
├── auth-istio
│   ├── ...
│   ├── docs/
│   └── README.md // Current readme file for Istio based implementation
├── auth-apisix
│   ├── ...
│   └── README.md // readme file for APISIX based implementation
├── auth-oidc-provider
│   └── keycloak_install.yaml // Or even use the community installation method and note that in both readmes: 
│                                              // kubectl create -f https://raw.githubusercontent.com/keycloak/keycloak- 
│                                              // quickstarts/latest/kubernetes/keycloak.yaml
└── README.md // Overall readme file for authentication and authorization and links to different implementations
```
@lianhao @mkbhanda Kindly share your thoughts here.

Reply:
@Ruoyu-y how about we keep the keycloak installation separate in each different scenario, because I see the installation options of keycloak for istio-gateway and apisix is a little bit different. Also we can't make sure that in the future a 3rd auth scenario would use the same keycloak as well.

Reply:
> @Ruoyu-y how about we keep the keycloak installation separate in each different scenario, because I see the installation options of keycloak for istio-gateway and apisix is a little bit different. Also we can't make sure that in the future a 3rd auth scenario would use the same keycloak as well.

Sure. Let's put keycloak installation and guide in each directory.

Reply:
BTW, i just created a draft PR #353 to restructure the directory. @ckhened please check if possible to add apisix implementation under such structure.

Reply:
@Ruoyu-y with the directory structure in your PR, apisix can be added without any problem. I'll wait for your PR to be merged and then add my changes in the authN-authZ directory. Meanwhile i'll fix all the review comments

Reply:
Moved auth-apisix to new directory structure